### PR TITLE
[SPARK-14088][SQL] Some Dataset API touch-up

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
1. Deprecated unionAll. It is pretty confusing to have both "union" and "unionAll" when the two do the same thing in Spark but are different in SQL.
2. Rename reduce in KeyValueGroupedDataset to reduceGroups so it is more consistent with rest of the functions in KeyValueGroupedDataset. Also makes it more obvious what "reduce" and "reduceGroups" mean. Previously it was confusing because it could be reducing a Dataset, or just reducing groups.
3. Added a "name" function, which is more natural to name columns than "as" for non-SQL users.
4. Remove "subtract" function since it is just an alias for "except".
## How was this patch tested?

All changes should be covered by existing tests. Also added couple test cases to cover "name".

### CODE REVIEW

**Comment by rxin** [comment]
cc @davies for Python, and @cloud-fan / @liancheng for the rest.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 1]
```diff
@@ -937,6 +932,26 @@ class Column(protected[sql] val expr: Expression) extends Logging {
...
 
   /**
+   * Gives the column a name (alias).
+   * {{{
+   *   // Renames colA to colB in select output.
+   *   df.select($"colA".name("colB"))
+   * }}}
+   *
+   * If the current column has metadata associated with it, this metadata will be propagated
+   * to the new column.  If this not desired, use `as` with explicitly empty metadata.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Is there any way to avoid duplicating documents at scala side?

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: no unfortunately...

**Comment by davies** [comment]
LGTM for python part

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Well, actually there is. You can use `@define <name> <definition>` and then reference it using `$name` later. See [here][1]. Scala standard library uses this trick extensively.

[1]: http://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: does it work with the unidoc for java?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Just checked, unfortunately it doesn't work for Java :(

**Pull request closed without merging**